aten/src/ATen/native/RangeFactories.cpp

```diff
@@ -21,14 +22,14 @@ Tensor& linspace_cpu_out(Tensor& result, Scalar start, Scalar end, int64_t steps
   } else if (steps == 1) {
     r.fill_(start);
   } else {
-    AT_DISPATCH_FLOATING_TYPES(r.scalar_type(), "linspace_cpu", [&]() {
+    AT_DISPATCH_FLOATING_TYPES_AND_COMPLEX(r.scalar_type(), "linspace_cpu", [&]() {
       scalar_t scalar_start = start.to<scalar_t>();
       scalar_t scalar_end = end.to<scalar_t>();
       scalar_t *data_ptr = r.data_ptr<scalar_t>();
       scalar_t step = (scalar_end - scalar_start) / static_cast<scalar_t>(steps - 1);
       at::parallel_for(0, steps, internal::GRAIN_SIZE, [&](int64_t p_begin, int64_t p_end) {
         scalar_t is = static_cast<scalar_t>(p_begin);
-        for (int64_t i = p_begin; i < p_end; ++i, ++is) {
+        for (int64_t i = p_begin; i < p_end; ++i, is+=1) {
```

Comment:
What's going on with the increment here? lol

Reply:
`scalar_t is = static_cast<scalar_t>(p_begin);` ensures that `is` is of type `std::complex<double>` or `std::complex<float>`. These data types have very few operator overloads. See [std::complex](https://en.cppreference.com/w/cpp/numeric/complex).

These operators in question are (>, >=, <=, <, ++, --).

**I'm not sure what to do and could use your advice:**
1. Do we create a subclass of std:complex that includes more operators?
    - Note: std:: containers do not support polymorphism.
2. Do we just keep going and make fixes like this as needed?
   - Supposedly there is no difference in `++is` and `is+1` in modern compilers.
   - I was roughly able to implement all at::native functions (still missing some _th_ functions).
   - I have to add custom `DISPATCH` switch statements to avoid AVX/SSE calls anyways.
3. **I have the following code that can `DISPATCH` the sub stl-container type.**

```c++
#define AT_PRIVATE_CASE_SUB_TYPE(enum_type, type, sub_enum, sub_type, ...)       \
  case enum_type: {                                      \
    using scalar_t = type;                               \
    ScalarType scalar_sub_type = sub_enum;           \
    using scalar_sub_t = sub_type;                       \
    return __VA_ARGS__();                                \
  }

#define AT_DISPATCH_FLOATING_AND_COMPLEX_SUB_TYPES(TYPE, NAME, ...)                                              \
  [&] {                                                                                                          \
    const auto& the_type = TYPE;                                                                                 \
    /* don't use TYPE again in case it is an expensive or side-effect op */                                      \
    at::ScalarType _st = ::detail::scalar_type(the_type);                                                        \
    switch (_st) {                                                                                               \
      AT_PRIVATE_CASE_SUB_TYPE(at::ScalarType::Double, double, at::ScalarType::Double, double, __VA_ARGS__)                              \
      AT_PRIVATE_CASE_SUB_TYPE(at::ScalarType::Float, float, at::ScalarType::Float, float, __VA_ARGS__)                                 \
      AT_PRIVATE_CASE_SUB_TYPE(at::ScalarType::ComplexDouble, std::complex<double>, at::ScalarType::Double, double, __VA_ARGS__)         \
      AT_PRIVATE_CASE_SUB_TYPE(at::ScalarType::ComplexFloat, std::complex<float>, at::ScalarType::Float, float, __VA_ARGS__)            \
      default:                                                                                                   \
        AT_ERROR(#NAME, " not implemented for '", toString(_st), "'");                                           \
    }                                                                                                            \
  }()
```

I'm thinking of using the third approach.  I've  already run into some Intel MKL functions where this was needed. What are your thoughts?

Reply:
It seems useful to have the ability to get the "true" scalar type from a complex scalar type (which isn't really a scalar at all). But you should do this via a template that matches specifically on `std::complex<T>` and returns `T`. That way no need to change the dispatch macros.

Reply:
This would be analogous to `acctype_t`

Reply:
But in this particular case, I think `is+=1` is probably better, because it prevents us from having to copy paste the kernel.